### DISCUSSION TITLE
Provide more documentation in JSDoc comments

### DISCUSSION THREAD

**Discussion by jcready** [TanStack/query#5621] [Ideas]
Certainly there are already a lot of JSDoc comments in the code base providing helpful documentation especially handy inside IDEs that surface those. But there are also a lot of things that **don't** have JSDoc comments where they would definitely be useful.

For example, compare all the helpful information in the [`useQuery` docs for all the query options](https://tanstack.com/query/latest/docs/react/reference/useQuery) to the JSDocs available for the query options: https://github.com/TanStack/query/blob/3643b87c34d86117b150ca845a058991ff132ab0/packages/query-core/src/types.ts#L70-L80

Even less information is available when comparing the docs for the return type to what is available in the JSDoc comments: https://github.com/TanStack/query/blob/3643b87c34d86117b150ca845a058991ff132ab0/packages/query-core/src/types.ts#L374-L402

This means I often have to leave my IDE, open up the docs website, make sure I switch to the correct version of the docs, then find the documentation for the thing I'm trying to use. In my ideal world the JSDoc comments would serve as a sort of source of truth from which the documentation website would be derived, but that's probably a big ask.

Instead I'd just ask that more of the information available in the docs could be also made available in JSDoc comments and/or have JSDoc comments linking back to the official docs:

```ts
/**
 * @deprecated - use `structuralSharing` instead
 * @see https://tanstack.com/query/v4/docs/react/reference/useQuery#options-isDataEqual
 */
isDataEqual?: (oldData: TData | undefined, newData: TData) => boolean
```

Side request: provide more anchors/ids for parts of the docs so we can link directly to stuff like the `QueryOptions.isDataEqual` instead of only being able to link to `useQuery` and then ask the person to search for "isDataEqual".

**Comment by TkDodo** [comment]
yeah absolutely. If you want, please contribute this?
